In [2]:
from sklearn import datasets
import pandas as pd

boston = datasets.load_boston()

df = pd.DataFrame(boston.data, columns = boston.feature_names)
df["MEDV"] = boston.target
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

### 訓練用データ・テスト用データに分割する

In [3]:
# 変数xに犯罪率CRIM,平均部屋数RM,低所得者割合LSTAT
# 変数yに予想する住宅価格MEDV

x = df.loc[:, ["CRIM", "RM", "LSTAT"]]
y = df.loc[:, "MEDV"]

In [4]:
# train_test_splitで分割する
# 訓練用データ8割, テスト用データ2割とする

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, train_size=0.8)

In [5]:
print(len(x_train),len(x_test),len(y_train),len(y_test))

404 102 404 102


全データ数506個を訓練用データ402個,　テスト用データ102個に分割。

### 住宅価格予想のため回帰分析を行う

x軸パラメータ1つ・y軸パラメータ1つのものを単回帰分析

x軸パラメータ複数・y軸パラメータ1つのものを重回帰分析

回帰式が直線：線形回帰

回帰式が曲線：非線形回帰

今回は、犯罪率CRIM,平均部屋数RM,低所得者割合LSTATで重回帰分析の線形回帰。

In [6]:
# モデルインポート
from sklearn.linear_model import LinearRegression

In [7]:
model = LinearRegression()
model.fit(x_train, y_train)

LinearRegression()

これでモデルが決定、機械学習ができる状態に。　状態を確認すると・・・

In [8]:
y_pred = model.predict(x_test)

print(y_pred)

[17.2162117  25.98642623 28.76361572 36.5503584  19.98641573 31.41112032
 20.19751144 30.2384599  33.78232073 23.37360005 36.70159376 19.19788329
 25.03075702 38.59324524  6.40118757 21.46442941 19.22243051 35.88659044
  9.50335123  9.81976438 20.87287145 31.48939066 17.28905152 26.04799685
 21.93773275 23.21174916 20.46883738 35.32621558 18.88011492 21.02174321
 27.25598618 27.49010255 23.8856928  12.07241204 20.77636099 21.56662771
 28.99722898 22.36085061 30.80358552 26.75133668 28.03335338 21.07127715
 19.53885378 25.75968252 23.98362051 30.48253386 26.66937543  4.18410221
 32.3507988  37.37303616 16.51408259 26.50469581 22.04242646  9.28655739
 22.06602179 23.05289578 27.53901867 22.69937793 19.11569052 26.82567495
 21.68568086 24.26476539 31.85070884 37.93284602 22.89184732 25.07925789
 12.4795166  26.19619825 26.16972822 23.56355817 32.42748511 24.97347306
 15.68738565 18.16347532 30.16269812 20.91844059 28.96469885 15.39174208
 19.48212778 27.39045853 29.0306663  25.85367493 32

上記数字が、各データから予想される住宅価格。

線形回帰モデル評価指標に、決定係数R^2がある。

これは0～1の値を取り、1に近いほど機械学習の予想が的中していることを表す。

そして、r2_score（予想）として決定係数を計算する。

In [9]:
from sklearn.metrics import r2_score

In [10]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

model = LinearRegression()
model.fit(x_train, y_train)

y_pred = model.predict(x_test)
r2_score(y_test, y_pred)

0.6160085822150296

これではなんともいえない精度のため、より関連の高そうな特徴量パラメータを用いて行うことにする。

### 整理

犯罪率CRIM,平均部屋数RM,低所得者割合LSTATを変数xに、予想したい住宅価格MEDVを変数yに格納した。

更に　訓練用データに8割、テスト用データに2割として分割した。

今回は25000平方フィート以上の広い住宅区画割合ZN,小売業以外のビジネスがされている土地割合INDUS,窒素酸化物濃度NOXを変数に追加する。

これを訓練用データ,テスト用データに分割した。

In [14]:
x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)

変数x1,x2に機械学習評価を実施する

決定係数R^2のため、1に近いほど予想が正しいことを表す。

In [18]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

model1 = LinearRegression()
model2 = LinearRegression()

model1.fit(x1_train, y1_train)
model2.fit(x2_train, y2_train)

y1_pred = model1.predict(x1_test)
y2_pred = model2.predict(x2_test)

print(r2_score(y1_test, y1_pred), r2_score(y2_test, y2_pred))

0.656352190695449 0.6428198564534187


3種のデータによる決定係数は0.656352190695449,6種のデータによる決定係数は0.6428198564534187である。

そこでより正しい判断をしたい。

### for関数による繰り返しデータ分割・モデル作成と機械学習・予想と評価を実施してスコア平均を表示する

1回目

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6351872750697238 0.6199228480099284


2回目

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.625142483393305 0.631089837543912


3回目

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6325696090428147 0.6356923687751794


4回目

In [23]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.636412642761546 0.6197549895256084


5回目

In [24]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6260669379656322 0.6209666183244829


6回目

In [25]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6339309085996365 0.635757946383764


7回目

In [26]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6409536162441724 0.6255186668598682


8回目

In [27]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6296574241437922 0.6236397529064652


9回目

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6346777453533405 0.633305767242345


10回目

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np

trial = 100

x1 = df.loc[:, ["CRIM", "RM", "LSTAT"]]
x2 = df.loc[:, ["CRIM", "RM", "LSTAT", "ZN", "INDUS", "NOX"]]
y = df.loc[:, "MEDV"]

pred1_score=[]; pred2_score=[]

for i in range(trial):
    x1_train, x1_test, y1_train, y1_test= train_test_split(x1, y, test_size=0.2, train_size=0.8)
    x2_train, x2_test, y2_train, y2_test= train_test_split(x2, y, test_size=0.2, train_size=0.8)
    
    model1 = LinearRegression()
    model2 = LinearRegression()
    
    model1.fit(x1_train, y1_train)
    model2.fit(x2_train, y2_train)

    y1_pred = model1.predict(x1_test)
    y2_pred = model2.predict(x2_test)
    
    pred1_score.append(r2_score(y1_test, y1_pred))
    pred2_score.append(r2_score(y2_test, y2_pred))
    
pred1_ave = np.average(np.array(pred1_score))
pred2_ave = np.average(np.array(pred2_score))

print(pred1_ave, pred2_ave)

0.6233222293358687 0.6240965801400692


今回はひとまず100回の試行で平均値を10回求めてみた。

データ種類

                3種類　　　　　　　　　　　　 6種類

                0.6351872750697238            0.6199228480099284
                
                0.625142483393305             0.631089837543912
                
                0.6325696090428147            0.6356923687751794
                
                0.6260669379656322            0.6209666183244829
                
                0.6260669379656322            0.6209666183244829
                
                0.6339309085996365            0.635757946383764
                
                0.6409536162441724            0.6255186668598682
                
                0.6296574241437922            0.6236397529064652
                
                0.6346777453533405            0.633305767242345
                
                0.6233222293358687            0.6240965801400692

平均値は　　　　3種類：0.630757517　　　　　　6種類：0.6270957

検証したデータ要素の多い6要素のほうがスコアが低くなっている。

つまり、追加した25000平方フィート以上の広い住宅区画割合ZN,小売業以外のビジネスがされている土地割合INDUS,窒素酸化物濃度NOXは

ボストン住宅価格MEDVとは関係が低いことが判明した。